### ISSUE TITLE
no-useless-escape false negative

### ISSUE THREAD

**Issue by Trott** [eslint/eslint#7772]
**Tell us about your environment**

* **ESLint Version:** 3.12.2
* **Node Version:** 7.2.1
* **npm Version:** 3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**
n/a, no configuration file

```




```

**What did you do? Please include the actual source code causing the issue.**

Running `eslint --no-eslintrc --rule 'no-useless-escape: 2' test.js` with the contents of `test.js being:

```js
/\{/
```

Also same results with this more complete example:

```js
console.log(/\{/.test('{')); // true

console.log(/{/.test('{')); // also true
```

**What did you expect to happen?**

I expected to be told that `{` was being uselessly escaped on line 1. 

**What actually happened? Please include the actual, raw output from ESLint.**

No error reported. This issue also applies to `/\}/`.

**Comment by not-an-aardvark** [comment]
I was able to reproduce this, thanks for the report.

These are a few relevant test cases:

```js
/\{/; // useless
/\}/; // useless
/\{1}/; // not useless (SyntaxError if removed)
/{1/}/; // not useless (SyntaxError if removed)
/foo\{1}/; // not useless
/foo{1\}/; // not useless
/foo\{a}/; // useless (this matches the string 'foo{a}')
/foo{a\}/; // useless
/\{{2}/; // useless (this matches the character '{' twice
/{\{2}/; // not useless
```

I'm not sure I fully understand the parsing rules for cases where there is something other than a number in curly brackets. I'll have to look into it further.

**Comment by not-an-aardvark** [comment]
Based on the [spec](https://tc39.github.io/ecma262/#prod-QuantifierPrefix), it looks like curly brackets are only considered part of quantifiers if they are not in a character class, and they match one of the following patterns:

```
{digits}
{digits,}
{digits,digits}
```

Aside from the question of whether the braces are in a character class (which we already detect elsewhere in the rule), I don't think this check is context-dependent, so any escaped curly brace that's not part of a pattern like this should be reported.

**Comment by not-an-aardvark** [comment]
I'm wondering how to handle this case:

```js
/a\{1\}/;
```

The first escape is useless, because the regular expression is functionally identical to

```js
/a{1\}/;
```

Independently, the second escape is also useless, because the regular expression is functionally identical to

```js
/a\{1}/;
```

But it only makes sense to fix one of the escapes, because if both characters are unescaped, the resulting expression has a different meaning:

```js
/a{1}/;
```

I suppose it's technically correct to report both escapes, because as written, both of them are independently useless. But I'm worried that might be confusing if the user tries to fix both issues at once and accidentally changes the behavior of their regex. Maybe we should just pick one of the escapes to report here.

**Comment by not-an-aardvark** [comment]
Actually, after looking at the spec I'm not sure this is a bug -- see [this comment](https://github.com/eslint/eslint/issues/7656#issuecomment-267878150) for more explanation. `/{/` is only a legal regex for engines that implement [Annex B](https://tc39.github.io/ecma262/#sec-regular-expressions-patterns), a part of the spec which is only normative for engines in web browsers. So `/\{/` is only a useless escape on engines that implement Annex B, which might not apply to all engines.

**Comment by not-an-aardvark** [comment]
Closing this issue -- as described above, `/\{/` is only a useless escape for engines that implement Annex B, but it's not useless for JS engines in general.
